Fill out examples

fitzgen opened the pull request:
@alexcrichton want to take a look at the TodoMVC implementation?

If you're interested in some lolsob, also check out 5837d53...

fitzgen committed:
Add READMEs and comments for examples

fitzgen committed:
Move `Render` to a private submodule

fitzgen committed:
Deny missing docs and missing `Debug` implementations

fitzgen committed:
Only build examples as cdylib

fitzgen committed:
Fix a bug where "volatile" attributes weren't getting properly updated

fitzgen committed:
Add an example TodoMVC implementation

fitzgen committed:
Build with LTO

fitzgen committed:
Update the lockfile

alexcrichton commented on Cargo.toml:
```diff
@@ -41,12 +41,14 @@ features = [
 
 [profile.release]
-# Tell `rustc` to optimize for small code size.
+incremental = false
```
[Not for long](https://github.com/rust-lang/rust/pull/58378)!

alexcrichton commented on examples/todomvc/Cargo.toml:
```diff
@@ -0,0 +1,58 @@
+optional = true
+version = "0.4.6"
+
+[dependencies.serde]
```
I might recommend collapsing these using inline tables where possible (as it's a bit more idiomatic I think)

fitzgen replied:
I updated my `cargo update` but it still is canonicalizing everything like this :(

alexcrichton replied:
Yeah I don't think they're using something which preserves formatting (not sure if that exists in a production-ready capacity even...)

alexcrichton commented on examples/todomvc/Cargo.toml:
```diff
@@ -0,0 +1,58 @@
+
+[dependencies.log]
+optional = true
+version = "0.4.6"
```
You might be able to expose some features of this to compile out `debug!` and/or `info!` statements perhaps?

fitzgen replied:
This crate itself doesn't do any logging, just `dodrio` but it does need to depend on `log` to get the `Level` when initializing the `console_log` logger. So I think it is OK to tie them together in the `logging` feature since they are only ever used together.

alexcrichton replied:
Oh for sure, but `debug!`'s code is emitted by default, which means there's a runtime check and the code for formatting is all compiled in. For the tiniest binaries we probably want to compile out the log messages?

fitzgen replied:
I've been seeing some `debug!`s get compiled away when there is no logger -- should double check that for `dodrio` for sure though.

alexcrichton replied:
Oh right that's a good point. I think LTO is good enough to see that if no one sets the max log level the max log level is a constant that forbids logging so everything is optimized away as if it were already compiled out

alexcrichton commented on examples/todomvc/index.html:
```diff
@@ -0,0 +1,21 @@
+        <meta charset="utf-8">
+        <meta name="viewport" content="width=device-width, initial-scale=1">
+        <title>dodrio • TodoMVC</title>
+        <link rel="stylesheet" href="node_modules/todomvc-common/base.css">
```
Is there perhaps an NPM CDN thing we could use to skip the `npm install` step for this?

fitzgen replied:
Good call: https://unpkg.com/

alexcrichton commented on examples/todomvc/src/controller.rs:
```diff
@@ -0,0 +1,124 @@
+
+/// Invoke `f` with the root `Todos` component, save the (likely just modified)
+/// todos to local storage, and schedule a new `dodrio` render.
+fn with_todos<F>(root: &mut dyn RootRender, vdom: VdomWeak, f: F)
```
Using RAII may make this a bit more ergonomic to call by skipping some indentation maybe?

alexcrichton replied:
Not more ergonomic to write initially though

alexcrichton commented on examples/todomvc/src/lib.rs:
```diff
@@ -0,0 +1,59 @@
+    let document = utils::document();
+    let container = document
+        .query_selector(".todoapp")
+        .unwrap_throw()
```
Could one of these be skipped by returning `Result` on the outer `run`?

alexcrichton commented on examples/todomvc/src/lib.rs:
```diff
@@ -0,0 +1,59 @@
+}
+
+cfg_if::cfg_if! {
+    if #[cfg(feature = "logging")] {
```
Is there a way, with wasm-pack today, to enable this feature?

fitzgen replied:
Yes, you can do `wasm-pack build -- $FLAGS_PASSE_DIRECTLY_TO_CARGO` since 0.6.0

alexcrichton commented on examples/todomvc/src/router.rs:
```diff
@@ -0,0 +1,62 @@
+/// Start the router.
+pub fn start(vdom: VdomWeak) {
+    // Callback fired whenever the URL's hash fragment changes. Keeps the root
+    // TODOs collection's visibility in sync with the `#` fragment.
```
now this is tricky

It's a comment that starts with "TODO"

not a TODO comment

alexcrichton commented on examples/todomvc/src/router.rs:
```diff
@@ -0,0 +1,62 @@
+    // Call it once to handle the initial `#` fragment.
+    on_hash_change();
+
+    // Now listen for hash changes forever.
```
Perhaps a small comment could be added about how this technically isn't linking memory within the context of this one application?

alexcrichton commented on examples/todomvc/src/todo.rs:
```diff
@@ -0,0 +1,229 @@
+    }
+
+    /// Get this TODO's title.
+    pub fn title(&self) -> &str {
```
I smell a derive macro for stuff like this

fitzgen replied:
or just make it all `pub`... a real application would probably not use `pub` so I didn't here either. This is kind of an in-between scenario.

alexcrichton commented on examples/todomvc/src/todo.rs:
```diff
@@ -0,0 +1,229 @@
+        let title = self.edits.as_ref().unwrap_or(&self.title);
+        let elem_id = bumpalo::format!(in bump, "todo-{}", id);
+
+        let mut input_attrs = bumpalo::vec![
```
Just gotta say. This, ergonomically, is pretty awesome

alexcrichton commented on examples/todomvc/src/todo.rs:
```diff
@@ -0,0 +1,229 @@
+        }
+        let input_attrs = input_attrs.into_bump_slice();
+
+        Node::element(
```
Heh we should definitely clean this up a bit at some point :)

(the helper methods you were mentioning would be perfect)

(well, that plus maybe a builder to add attributes)

alexcrichton reviewed (approved):
This was quite a pleasure to read over! All the pieces really are seriously coming together here...

I do think we gotta put some elbow grease into creating nodes before this super-awesome, but I'll settle for just awesome for now

alexcrichton commented:
Also I demand validation-via-travis.

And I request publication-via-travis.

fitzgen committed:
Bump `bumpalo` dependency (ha)

fitzgen committed:
Use inline tables for dependencies in TodoMVC's Cargo.toml

fitzgen committed:
Use unpkg to get TodoMVC's CSS

We can skip an `npm install` now.

fitzgen committed:
TodoMVC: Use an RAII type to auto render and save to local storage

Instead of a `with<F>`-style function. It is a bit more idiomatic, and provides
nicer usage, but takes a little more to write.

fitzgen committed:
Make sure we specify the latest version of all dependencies in our `Cargo.toml`s

fitzgen committed:
TodoMVC: use `?` instead of unwrapping in `run` start function

fitzgen committed:
Use 'todo' instead of 'TODO' for comments in TodoMVC

So they don't look like things we need to fix...

fitzgen committed:
TodoMVC: add small note about why we `forget` in the router

fitzgen committed:
Fix doc tests
